### PR TITLE
Change ArtifactResponse extracting method

### PR DESCRIPTION
**Pull request by vihanga-liyanage**
### Proposed changes in this pull request
- Earlier a regex pattern was used to extract the ArtifactResponse object. Hardcoded regex string can cause problems when specification versions or lib versions are upgraded. Therefore we have used library constants to validate and extract the object.

Refers wso2/product-is#3401
Resolves https://github.com/wso2/product-is/issues/3580

### When should this PR be merged
ASAP

### Follow up actions
- Update the pom version.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/TestConstants.java [hunk 1]
```diff
@@ -353,6 +353,77 @@ private TestConstants() {
 
     public static final String SAML_ART = "AAQAAM1OZiCJt3Va5A1W4dyuSepX9Q6clxUTA4vGKbi/bFCqu6Vm+NTZMDE=";
 
+    public static final String SAML_ART_RESPONSE = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/TestConstants.java]: Seems like this is not used in the testcases. If so, shall we remove this?

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionServiceTest.java [hunk 2]
```diff
@@ -36,10 +37,12 @@
 import java.util.Map;
 
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
 
 public class SAMLSSOArtifactResolutionServiceTest {
 
     private static ArtifactResolve artifactResolve;
+    private static ArtifactResponse artifactResponse;
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionServiceTest.java]: No need of these changes in this class.

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java [hunk 3]
```diff
@@ -116,26 +132,55 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
...
-            if (log.isDebugEnabled()) {
-                log.debug("Artifact Response: " + artifactResponseReceived);
-            }
-            try {
-                ArtifactResponse artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(artifactResponseReceived);
-                if (isArtifactResponseValid(artifactResolve, artifactResponse)) {
-                    return artifactResponse;
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
+        if (artifactResponse != null && isArtifactResponseValid(artifactResolve, artifactResponse)) {
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java]: Null check is already added inside isArtifactResponseValid() [1], no need to add it in here.

[1] https://github.com/vihanga-liyanage/identity-outbound-auth-samlsso/blob/2cf2a4fe6423a3c1aa3348a8401f9100c3bfb2ca/components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java#L189

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java [hunk 4]
```diff
@@ -116,26 +132,55 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
...
+     *
+     * @param artifactResponseString Response string from artifact resolver.
+     * @return Extracted artifact response object.
+     * @throws ArtifactResolutionException
+     */
+    public ArtifactResponse extractArtifactResponse(String artifactResponseString)
+            throws ArtifactResolutionException {
+
+        ArtifactResponse artifactResponse = null;
+        InputStream stream = new ByteArrayInputStream(artifactResponseString.getBytes(StandardCharsets.UTF_8));
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java:152]: Initially validate whether artifactResponseString is null.

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java:152]: Validation added before calling the method.

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java [hunk 5]
```diff
@@ -116,29 +132,59 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
...
-            try {
-                ArtifactResponse artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(artifactResponseReceived);
-                if (isArtifactResponseValid(artifactResolve, artifactResponse)) {
-                    return artifactResponse;
+        if (artifactResponseString == null) {
+            throw new ArtifactResolutionException("SOAP response was null when resolving artifact: " +
+            artifactResolve.getArtifact().getArtifact() + " Issuer: " + artifactResolve.getIssuer().getValue());
+        }
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java]: Remove the unnecessary line.

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java [hunk 6]
```diff
@@ -116,29 +132,59 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
...
-        if (matcher.find()) {
-            String artifactResponseReceived = matcher.group(0);
-            if (log.isDebugEnabled()) {
-                log.debug("Artifact Response: " + artifactResponseReceived);
-            }
-            try {
-                ArtifactResponse artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(artifactResponseReceived);
-                if (isArtifactResponseValid(artifactResolve, artifactResponse)) {
-                    return artifactResponse;
+        if (artifactResponseString == null) {
```
Comment (malithie) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java]: Is there any probability for this string to be empty

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java [hunk 7]
```diff
@@ -116,29 +132,59 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
...
+
+                    DOMSource source = new DOMSource(artifactResponseElement);
+                    StringWriter stringResult = new StringWriter();
+                    TransformerFactory.newInstance().newTransformer().transform(
+                            source, new StreamResult(stringResult));
+                    artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(stringResult.toString());
                 } else {
-                    throw new ArtifactResolutionException("Artifact Response is not valid.");
+                    throw new ArtifactResolutionException("Didn't receive valid artifact response.");
                 }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java]: So based on this if else block if the very first element is not an artifact response an error will be thown.
Is this the expectation. Can artifact response be nested ?

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java]: According to the spec, the SOAP message should wrap the Artifact Response object. Following is an example given in the spec.

```
<SOAP-ENV:Envelope
    xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
    <SOAP-ENV:Body>
        <samlp:ArtifactResponse
            xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
            xmlns="urn:oasis:names:tc:SAML:2.0:assertion"
            ID="_FQvGknDfws2Z" Version="2.0"
            InResponseTo="_6c3a4f8b9c2d"
            IssueInstant="2004-01-21T19:00:49Z">
            <Issuer>https://wso2is.com</Issuer>
            <samlp:Status>
                <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
            </samlp:Status>
            <samlp:Response ID="d2b7c388cec36fa7c39c28fd298644a8"
                IssueInstant="2004-01-21T19:00:49Z"
                Version="2.0">
                ...
            </samlp:Response>
        </samlp:ArtifactResponse>
    </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

**Review by malithie — APPROVED** [review]

**Pull request merged**